src/WP_CLI/SearchReplacer.php

```diff
@@ -123,10 +131,28 @@ private function _run( $data, $serialised, $recursion_level = 0, $visited_data =
 							)
 						);
 					}
-					$data = $result;
+				} elseif ( $this->callback ) {
+					if ( strpos( $data, $this->from ) !== false ) {
+						$result = \call_user_func( $this->callback, $data, $this->to );
+					} else {
+						// We can skip calling the function here. It must still be set so we don't remove text.
+						$result = $data;
+					}
 				} else {
-					$data = str_replace( $this->from, $this->to, $data );
+					$result = str_replace( $this->from, $this->to, $data );
+				}
+
+				if ( $this->callback ) {
+					if ( false === $result || is_wp_error( $result ) ) {
+						\WP_CLI::warning( 'The callback function threw an error. Stopping operation.' );
+						exit;
+					} elseif ( ! is_string( $result ) ) {
+						\WP_CLI::warning( 'The callback function did not return a string. Stopping operation.' );
```

Comment:
This should be an error, not a warning. This also removes the need for the exit in the next line, as the error will include halting the process.
```suggestion
						\WP_CLI::error( 'The callback function did not return a string. Stopping operation.' );
```